### PR TITLE
Pagination Component

### PR DESCRIPTION
**Pull request by paultannenbaum**
- Creation of pagination component. The component figures out how many pagination links to display based off the chunk size and the total list size. It handles the UI interactions that happen within the component, but because it is doing server side pagination, it sends its actions up to its controller allowing the data requesting and setting to happen at that level.
- Altered the getKeyList and getBuckets api calls to allow start and rows
- Set default values (currently 100)

Right now it has "Prev" and "Next" buttons prepending and appending the pagination link list, but these can easily be removed or have extra logic added on when to show them. Just wanted to have them in there in as an option, will prob decide later whats the best aesthetically.

### CODE REVIEW

File: app/pods/bucket-type/template.hbs [hunk 1]
```diff
@@ -145,15 +145,17 @@
...
-              <ul class='button-list small'>
-                {{#each model.bucketList.buckets as |bucket|}}
-                  <li>
-                    {{link.link-bucket bucket=bucket}}
-                  </li>
-                {{else}}
-                  No buckets found
-                {{/each}}
-              </ul>
+              {{#pagination-component chunk-length=1 total-length=model.bucketList.total section-request='retrieveRequestedBuckets'}}
```
Comment (paultannenbaum) [inline app/pods/bucket-type/template.hbs]: I noticed that the bucketlist model has two properties, `total` and `count`. What is the difference? With my seed data these two values are different, with `total` reflecting all list items. I'm not sure what `count` represents.

File: app/services/explorer.js [hunk 2]
```diff
@@ -655,44 +655,41 @@ export default Ember.Service.extend({
...
-        var clusterId = cluster.get('clusterId');
-        var bucketTypeId = bucketType.get('bucketTypeId');
-        var url = this.apiURL + 'explore/clusters/' + clusterId +
-            '/bucket_types/' + bucketTypeId + '/buckets' ;
-        var explorer = this;
+    getBucketList(cluster, bucketType, store, start=1, rows=1) {
+        let explorer = this;
+        let clusterId = cluster.get('clusterId');
+        let bucketTypeId = bucketType.get('bucketTypeId');
+        let url = `${this.apiURL}explore/clusters/${clusterId}/bucket_types/${bucketTypeId}/buckets?start=${start}&rows=${rows}`;
```
Comment (paultannenbaum) [inline app/services/explorer.js]: FYI, I cleaned this up to use [ES6 string interpolation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/template_strings)

I set the `start` and `rows` defaults manually ([another ES6 feature](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Default_parameters):smile:), but I am curious where you think these should live? `config/environment` doesn't make sense. Maybe here is fine?

File: app/services/explorer.js [hunk 3]
```diff
@@ -760,12 +757,14 @@ export default Ember.Service.extend({
...
-                return bucketType;
-            });
+    // TODO: What is the difference between this and getBucketList???
+    getBucketTypeWithBucketList(bucketType, cluster, store, start, row) {
+        return this
+          .getBucketList(cluster, bucketType, store, start, row)
+          .then(function(bucketList) {
+              bucketType.set('bucketList', bucketList);
+              return bucketType;
+          });
```
Comment (paultannenbaum) [inline app/services/explorer.js]: I'm not seeing the difference between this and `getBucketList`. Would you mind explaining how they are different.

File: app/services/explorer.js [hunk 4]
```diff
@@ -893,39 +894,38 @@ export default Ember.Service.extend({
      * @param {DS.Store} store
      * @return {Ember.RSVP.Promise} result of the AJAX call
      */
-    getKeyList(bucket, store) {
-        var clusterId = bucket.get('clusterId');
-        var bucketTypeId = bucket.get('bucketTypeId');
-        var bucketId = bucket.get('bucketId');
-        var explorer = this;
+    getKeyList(bucket, store, start=1, rows=100) {
```
Comment (paultannenbaum) [inline app/services/explorer.js]: Same goes for this one. Alter the default value in `rows` if you want to test this out locally and want the pagination chunks to be shorter.

File: app/services/explorer.js [hunk 5]
```diff
@@ -655,44 +657,42 @@ export default Ember.Service.extend({
...
      * @return {Ember.RSVP.Promise<BucketList>} Result of the AJAX request
      */
-    getBucketList(cluster, bucketType, store) {
-        console.log('Refreshing buckets for bucketType: %O', bucketType);
-        var clusterId = cluster.get('clusterId');
-        var bucketTypeId = bucketType.get('bucketTypeId');
-        var url = this.apiURL + 'explore/clusters/' + clusterId +
-            '/bucket_types/' + bucketTypeId + '/buckets' ;
-        var explorer = this;
+    getBucketList(cluster, bucketType, store, start=1, rows=100) {
```
Comment (paultannenbaum) [inline app/services/explorer.js]: Alter the default value in rows if you want to test this out locally and want the pagination chunks to be shorter. This value gets set as an attribute on the model, and the component picks it up from there.

File: app/services/explorer.js [hunk 6]
```diff
@@ -655,44 +657,42 @@ export default Ember.Service.extend({
...
-                    var emptyList = store.createRecord('bucket-list', {
-                        cluster: cluster,
-                        bucketType: bucketType
-                    });
-                    Ember.run(null, resolve, emptyList);
-                } else {
-                    Ember.run(null, reject, textStatus);
+                type: 'GET',
+                success: function(data) {
+                    data.maxBucketsPerRequest = rows; // Pass through what the chunk size should be
```
Comment (paultannenbaum) [inline app/services/explorer.js]: This is how the default value that is set up above is passed through, it is attached to the ajax request data.

**Comment by dmitrizagidulin** [comment]
> I noticed that the bucketlist model has two properties, total and count. What is the difference? With my seed data these two values are different, with total reflecting all list items. I'm not sure what count represents.

@paultannenbaum - Total is 'Total number of items in the cached (key or bucket) list', Count is 'Number of items displayed on the current page of the list'. Take a look at the `CachedList` models both the bucket list and key list inherit from.

**Comment by dmitrizagidulin** [comment]
> I'm not seeing the difference between `getBucketTypeWithBucketList` and `getBucketList`.

From what I recall, this structure came about when I was still learning / struggling with Ember Promises, the route's `model()` function, and the async nature of loading the bucket list. 
Let me see if I can refactor it..

File: app/models/bucket-list.js [hunk 7]
```diff
@@ -37,6 +37,14 @@ var BucketList = CachedList.extend({
     cluster: DS.belongsTo('cluster'),
 
     /**
+     * The max amount of records that should be fetched in a given request
+     * @property maxBucketsPerRequest
+     * @type Integer
+     * @default 100
+     */
+    maxBucketsPerRequest: DS.attr('integer', { defaultValue: 100 }),
```
Comment (paultannenbaum) [inline app/models/bucket-list.js]: This should be put on inherited cached list model. Will update with rest of feedback.

File: app/models/key-list.js [hunk 8]
```diff
@@ -26,6 +26,14 @@ var KeyList = CachedList.extend({
     cluster: DS.attr(),
 
     /**
+     * The max amount of records that should be fetched in a given request
+     * @property maxKeysPerRequest
+     * @type Integer
+     * @default 100
+     */
+    maxKeysPerRequest: DS.attr('integer', { defaultValue: 100 }),
```
Comment (paultannenbaum) [inline app/models/key-list.js]: This should be put on inherited cached list model. Will update with rest of feedback.

**Comment by dmitrizagidulin** [comment]
+1, looks great.

**Pull request merged**